url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/cores/Core.java

```diff
@@ -0,0 +1,414 @@
+                    // in that turn, and if there was a sole element left
+                    
+                    // This is the case when we processed something (lastHandledI > 0)
+                    // and when element 2nd element to the right (the next one) exists
+                    // but its pair (the 3rd one) does not. 
+                    
+                    // In that case we connect this element wit the next one 
+                    
+                    if (lastHandledI + 2 * elementDistance <= upperBound && 
+                        	lastHandledI + 3 * elementDistance > upperBound &&
+                        	lastHandledI >= 0) {
+                        	folder.handle(lastHandledI, lastHandledI + 2 * elementDistance, lastHandledI);
+                    }
+                
+                    // At this point we finished the round
+                    try {
+                        levelbarrier.await();
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    } catch (BrokenBarrierException e) {
+                        e.printStackTrace();
+                    }
+                    
+                    
+                    // If we were the one who changed the level, we also change the baseCount 
+                    // back to 0
+                    if (round.compareAndSet(rnd, rnd + 1)) {
+                        baseCount.set(0);
+                    }
+                    
+                   
+                    // We first have to wait twice, as otherwise an other thread
+                    // might be quicker than us in returning up to the loop again and use the
+                    // old level value 
+                    
+
+                    // Now we all wait that the level has been updated
+                    try {
+                        levelbarrier.await();
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    } catch (BrokenBarrierException e) {
+                        e.printStackTrace();
+                    }
+                    
+                    // And get new parameters
+                    rnd = round.get();
+                    elementDistance = (int) Math.pow(2, rnd);
+                }
+
+                try {
+                    barrier.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                } catch (BrokenBarrierException e) {
+                    e.printStackTrace();
+                }
+            }
+        };
+
+        // Execute all our runnables.
+        for(int i=0; i<NUM_THREADS; i++) {
+            this.commonCore.executor().getExecutor().execute(runner);
+        }
+
+        // Wait for all threads to finish ...
+        try {
+            barrier.await();
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)